File: tianshu/core/llm/factory.py
from __future__ import annotations

from tianshu.core.llm.base import BaseProvider
from tianshu.core.llm.openai import OpenAIProvider

_PROVIDER_TYPES: dict[str, type[BaseProvider]] = {
    "openai": OpenAIProvider,
    "ollama": OpenAIProvider,
    "deepseek": OpenAIProvider,
    "qwen": OpenAIProvider,
    "moonshot": OpenAIProvider,
    "zhipu": OpenAIProvider,
    "kimi": OpenAIProvider,
    "minimax": OpenAIProvider,
    "groq": OpenAIProvider,
    "openrouter": OpenAIProvider,
    "together": OpenAIProvider,
    "siliconflow": OpenAIProvider,
    "fireworks": OpenAIProvider,
    "mistral": OpenAIProvider,
    "lingyiwanwu": OpenAIProvider,
    "stepfun": OpenAIProvider,
    "baichuan": OpenAIProvider,
    "volcengine": OpenAIProvider,
    "hunyuan": OpenAIProvider,
    "nvidia": OpenAIProvider,
    "perplexity": OpenAIProvider,
    "cohere": OpenAIProvider,
    "together_ai": OpenAIProvider,
    "oneapi": OpenAIProvider,
    "newapi": OpenAIProvider,
    "litellm": OpenAIProvider,
    "vllm": OpenAIProvider,
    "sglang": OpenAIProvider,
    "lmstudio": OpenAIProvider,
    "custom": OpenAIProvider,
    "local": OpenAIProvider,
}

# 非 OpenAI 兼容协议,需要独立 Provider 实现(按需通过 register_provider 注册)


def register_provider(name: str, cls: type[BaseProvider]) -> None:
    _PROVIDER_TYPES[name] = cls


def create_provider(name: str, base_url: str, model: str, api_key: str = "", **kwargs) -> BaseProvider:
    cls = _PROVIDER_TYPES.get(name)
    if cls is None:
        cls = OpenAIProvider
        _PROVIDER_TYPES[name] = OpenAIProvider
    return cls(base_url=base_url, model=model, api_key=api_key, **kwargs)


def available_providers() -> list[str]:
    return sorted(_PROVIDER_TYPES)

File: tiankui-ts/src/core/llm/factory.ts
import { OpenAIProvider } from "./openai.js";
import type { BaseProvider } from "./types.js";

const _PROVIDER_TYPES = new Map<string, new (baseUrl: string, apiKey: string, model: string, temperature?: number, timeout?: number, maxTokens?: number | null) => BaseProvider>();

export interface ProviderOptions {
  base_url: string;
  api_key?: string;
  model?: string;
  temperature?: number;
  timeout?: number;
  max_tokens?: number | null;
}

for (const name of [
  "openai", "ollama", "deepseek", "qwen", "moonshot", "zhipu", "kimi", "minimax",
  "groq", "openrouter", "together", "siliconflow", "fireworks", "mistral",
  "lingyiwanwu", "stepfun", "baichuan", "volcengine", "hunyuan", "nvidia",
  "perplexity", "cohere", "together_ai", "oneapi", "newapi", "litellm",
  "vllm", "sglang", "lmstudio", "custom", "local",
]) {
  _PROVIDER_TYPES.set(name, OpenAIProvider as unknown as new (b: string, k: string, m: string, t?: number, to?: number, mt?: number | null) => BaseProvider);
}

type ProviderCtor = new (baseUrl: string, apiKey: string, model: string, temperature?: number, timeout?: number, maxTokens?: number | null) => BaseProvider;

export function registerProvider(name: string, cls: ProviderCtor): void {
  _PROVIDER_TYPES.set(name, cls);
}

export function createProvider(
  name: string,
  base_url: string,
  model: string,
  api_key = "",
  opts: Partial<ProviderOptions> = {},
): BaseProvider {
  let cls = _PROVIDER_TYPES.get(name);
  if (!cls) {
    cls = OpenAIProvider as unknown as new (b: string, k: string, m: string, t?: number, to?: number, mt?: number | null) => BaseProvider;
    _PROVIDER_TYPES.set(name, cls);
  }
  return new cls(base_url, api_key ?? "", model, opts.temperature ?? 0.2, opts.timeout ?? 120.0, opts.max_tokens ?? null);
}

export function availableProviders(): string[] {
  return [..._PROVIDER_TYPES.keys()].sort();
}
